Added new tasks form for running and queued experiments

sroderick-g5sro opened the pull request:
Fixes #207

I have created an initial write of the delete running experiments enhancement. I haven't finished the unit tests (because I need another day and also wanted to make sure I was on the right track first). I have used HTML forms rather than Django forms (forms.py) because it seemed easier to style these. I am also not sure about functionality for "admins" should they get a list of all running experiments. Currently I am only filtering on author. I haven't implemented a delete confirmation prompt or any send email functionality - can do if these are required.

sroderick-g5sro commented:
I wasn't sure how to distinquish between experiements run with different versions of models and protocols. For example... in the screen shot below:

![image](https://user-images.githubusercontent.com/3098326/65782795-5619c300-e146-11e9-92d1-d4879a957ff5.png)

both the "groupings" are for Model01/Protocol1 and they are only distinguished by their create date. In some cases this might be the same for different versions. Should I add a unique version identifier as is done below, on the models page

![image](https://user-images.githubusercontent.com/3098326/65783092-f1129d00-e146-11e9-93c1-304ea4b9fd2f.png)

jonc125 commented on weblab/experiments/views.py:
```diff
@@ -45,10 +47,33 @@ class ExperimentsView(TemplateView):
+    def get_queryset(self):
+        return self.model.objects.filter(
+            Q(author=self.request.user) &
+            Q(versions__status=ExperimentVersion.STATUS_QUEUED) |
```
Check precedence here - you might need extra brackets! (Probably better to have extra brackets for clarity anyway.)

sroderick-g5sro replied:
as suggested changed query to use running experiments

jonc125 commented on weblab/experiments/views.py:
```diff
@@ -391,6 +416,7 @@ def get_success_url(self, *args, **kwargs):
         return reverse('experiments:versions', args=[self.get_object().experiment.id])
 
 
+
```
Try to avoid adding extra blank lines :)

sroderick-g5sro replied:
noted - thanks

jonc125 commented on weblab/static/js/experiment_tasks.js:
```diff
@@ -0,0 +1,28 @@
+
```
You'll need to add a `require()` line for this in the top of `main.js` to get it made available.

sroderick-g5sro replied:
yes I have put this in now - thanks

jonc125 commented on weblab/templates/experiments/experiment_tasks.html:
```diff
@@ -0,0 +1,76 @@
+                                    <div class="suppl" id="taskVersionDetailBox" style="padding-bottom: 0px">
+                                        <small>Created&nbsp;</small>
+                                        <time>{{ version.created_at }}&nbsp;</time>
+                                        {% with version.files|length as numfiles %}
```
You probably don't need this level of detail!

sroderick-g5sro replied:
removed

jonc125 commented on weblab/templates/experiments/experiment_tasks.html:
```diff
@@ -0,0 +1,76 @@
+                    </small>
+                </div>
+                <ul id="taskList">
+                    {% for version in experiment.versions.all|dictsortreversed:"created_at" %}
```
Will need to check that previous completed versions of an experiment don't also show up here.

sroderick-g5sro replied:
I have done some tests and it seems to filter by currently running experiment

sroderick-g5sro replied:
now covered by unit tests - thanks

jonc125 commented:
> I have used HTML forms rather than Django forms (forms.py) because it seemed easier to style these.

Looks good to me!

> I am also not sure about functionality for "admins" should they get a list of all running experiments. Currently I am only filtering on author.

I think that's fine. Admins should be able to delete via the auto-generated admin interface anyway.

> I haven't implemented a delete confirmation prompt or any send email functionality - can do if these are required.

I don't think so - @MichaelClerx ?

Regarding the display, `Experiment` has a `def get_name(self, model_version=False, proto_version=False)` function which you can use to add the further detail. Or you might need to create a property version `full_name` or similar; I can't recall offhand how easy it is for a template to pass arguments to a member function!

However, an alternative approach would be to restructure the queries so your model is `RunningExperiment` (for quicker queries), and use annotate to add details, e.g.
```python
return RunningExperiment.objects.filter(
    Q(experiment_version__author=self.request.user)
).order_by(
    'experiment_version__created_at',
).annotate(
    author_name=F('experiment_version__author__full_name'),
    expt_name=F('experiment_version__experiment__name'),
).select_related(
    'experiment_version',
)
```

MichaelClerx commented:
> > I haven't implemented a delete confirmation prompt or any send email functionality - can do if these are required.
>
> I don't think so - @MichaelClerx ?

Agreed!

sroderick-g5sro commented:
@jonc125 Hi, I have completed the changes to the code and hopefully addresses some of the review comments. I haven't really got any tests for this change., also the CI is complaining about Import sort orders. I wonder if you could approve the code changes Ive done please and I will sort out the rest on Monday 16th when I return from leave

jonc125 commented on weblab/templates/experiments/experiment_tasks.html:
```diff
@@ -32,35 +32,18 @@ <h2>List of active tasks</h2>
+                            <input class="taskCancelCheckBox" type="checkbox" name="chkBoxes[]"
+                                   value="{{ running_experiment.experiment_version.id }}">
+                            <div class="experiment-{{ running_experiment.experiment_version.status }}" id="taskDetailBox">
+                               {{ running_experiment.experiment_version.experiment.name }} Version: {{ running_experiment.experiment_version.name }}
```
As a minor efficiency improvement, you can add the `annotate` I suggested for `expt_name` and use `running_experiment.expt_name` rather than `running_experiment.experiment_version.experiment.name` - it'll save a DB query per experiment_version.

sroderick-g5sro replied:
Thanks Jonathan. It didn't seem to let me return name using annotation. I got an error message with a list of supportable values none of which was name. I will try again though

jonc125 replied:
Ah, of course `Experiment.name` is a Python property not a DB field. Sorry!

Perhaps just `select_related` the experiment as well then?

sroderick-g5sro replied:
yes thanks, I have added that

jonc125 commented on weblab/experiments/views.py:
```diff
@@ -52,19 +49,21 @@ class ExperimentTasks(LoginRequiredMixin, ListView):
         """Process list of user selected checkboxes and kill running versions"""
         for ver in request.POST.getlist('chkBoxes[]'):
-            self.model.objects.filter(versions=ver).delete()
+            self.model.objects.filter(experiment_version_id=ver).delete()
```
I think this now wouldn't work, as self.model is `RunningExperiment` but the ids you're getting back are the `ExperimentVersion` ids. So probably need `ExperimentVersion.objects.filter`?

sroderick-g5sro replied:
thanks, It did seem to work ok.... Perhaps because the experiment version id field is also on the running experiemnts table? however your way is cleaner and more readable so I have implemented your suggestion here

jonc125 replied:
Indeed, because all id fields are just sequences of integers, I suspect in the test both happened to be 1!

jonc125 commented on weblab/static/js/main.js:
```diff
@@ -10,6 +10,7 @@ var experiment = require('./experiment.js');
 var notifications = require('./lib/notifications.js');
 require('./compare.js');
 require('./entity_version_list.js');
+require('./experiment_tasks');
```
I note all the others have `.js` appended - I'm presuming that this therefore isn't actually required?

sroderick-g5sro replied:
yes, good spot thanks. I have corrected the extension. Not sure why it was working locally without that - does it perhaps use the local .js files? rather than the gulped one?

sroderick-g5sro commented:
@jonc125 Hopefully I have covered all the review comments now. Looks like I still need to address the sort imports causing travis still.

jonc125 commented:
You can run `isort -rc .` on your machine to get isort to reorder imports automatically. The flake8 issues need to be addressed manually though.

jonc125 commented on weblab/experiments/tests/test_views.py:
```diff
@@ -892,6 +897,93 @@ def test_view_experiment_version(self, client, experiment_version):
+@pytest.mark.django_db
+class TestExperimentTasks:
+
+    @pytest.mark.usefixtures('other_user')
```
I don't think we need the fixture here; the client uses an anonymous user by default unless the `logged_in_user` fixture is used.

sroderick-g5sro replied:
thanks - changed

jonc125 commented on weblab/experiments/tests/test_views.py:
```diff
@@ -892,6 +897,93 @@ def test_view_experiment_version(self, client, experiment_version):
+        )
+
+        exp_version_2 = recipes.experiment_version.make(
+            status=ExperimentVersion.STATUS_RUNNING,
```
Actually, maybe make this one QUEUED to prove that both statuses count?

sroderick-g5sro replied:
good idea - done

jonc125 commented on weblab/experiments/tests/test_views.py:
```diff
@@ -892,6 +897,93 @@ def test_view_experiment_version(self, client, experiment_version):
+
+        # Return only running experiment versions
+        response = client.get('/experiments/tasks')
+        assert len(response.context['runningexperiment_list']) == 2
```
I wouldn't check both len and the set contents - checking the set contents will do both, and also give a more useful error message if the test fails (since you'll see which items didn't match).

sroderick-g5sro replied:
agreed

jonc125 commented on weblab/experiments/tests/test_views.py:
```diff
@@ -892,6 +897,93 @@ def test_view_experiment_version(self, client, experiment_version):
+        # Cancel one of the running versions check the other one is still present
+        client.post('/experiments/tasks', {'chkBoxes[]': [running_exp_version2.experiment_version.id]})
+        response = client.get('/experiments/tasks')
+        assert len(response.context['runningexperiment_list']) == 1
```
As above, delete this line.

sroderick-g5sro replied:
agreed

jonc125 commented on weblab/experiments/views.py:
```diff
@@ -45,6 +54,30 @@ class ExperimentsView(TemplateView):
+
+    def get_queryset(self):
+        return RunningExperiment.objects.filter(
+            Q(experiment_version__author=self.request.user)
```
You don't actually need the `Q()` wrapper for simple filtering cases like this; you can just do `.filter(experiment_version__author=self.request.user)`.

sroderick-g5sro replied:
good spot - done
